pennylane/pauli/pauli_arithmetic.py

```diff
@@ -414,10 +443,67 @@ def _pw_wires(w: Iterable) -> wires.Wires:
 
+    def _to_sparse_mat(self, wire_order, buffer_size=None):
+        """Compute the sparse matrix of the Pauli sentence by efficiently adding the Pauli words
+        that conform it. See pauli_sparse_matrices.md for the technical details."""
+        pauli_words = list(self)  # Ensure consistent ordering
+        n_wires = len(wire_order)
+        matrix_size = 2**n_wires
+        matrix = sparse.csr_matrix((matrix_size, matrix_size), dtype="complex128")
+        op_sparse_idx = _ps_to_sparse_index(pauli_words, wire_order)
+        _, unique_sparse_structures, unique_invs = np.unique(
+            op_sparse_idx, axis=0, return_index=True, return_inverse=True
+        )
+        pw_sparse_structures = unique_sparse_structures[unique_invs]
+
+        buffer_size = buffer_size or 2**30  # Default to 1GB of memory
+        # Convert bytes to number of matrices:
+        # complex128 (16) for each data entry and int64 (8) for each indices entry
+        buffer_size = max(1, buffer_size // ((16 + 8) * matrix_size))
+        mat_data = np.empty((matrix_size, buffer_size), dtype=np.complex128)
+        mat_indices = np.empty((matrix_size, buffer_size), dtype=np.int64)
+        n_matrices_in_buffer = 0
+        for sparse_structure in unique_sparse_structures:
+            indices, *_ = np.nonzero(pw_sparse_structures == sparse_structure)
+            mat = self._sum_same_structure_pws([pauli_words[i] for i in indices], wire_order)
+            mat_data[:, n_matrices_in_buffer] = mat.data
+            mat_indices[:, n_matrices_in_buffer] = mat.indices
+
+            n_matrices_in_buffer += 1
+            if n_matrices_in_buffer == buffer_size:
+                # Add partial results in batches to control the memory usage
+                matrix += self._sum_different_structure_pws(mat_indices, mat_data)
+                n_matrices_in_buffer = 0
+
+        matrix += self._sum_different_structure_pws(
+            mat_indices[:, :n_matrices_in_buffer], mat_data[:, :n_matrices_in_buffer]
+        )
+        return matrix
+
+    def _sum_same_structure_pws(self, pauli_words, wire_order):
+        """Sums Pauli words with the same sparse structure."""
+        mat = pauli_words[0].to_mat(wire_order, coeff=self[pauli_words[0]], format="csr")
+        for word in pauli_words[1:]:
+            mat.data += word.to_mat(wire_order, coeff=self[word], format="csr").data
+        return mat
+
+    @staticmethod
+    def _sum_different_structure_pws(indices, data):
+        """Sums Pauli words with different parse structures."""
+        size = indices.shape[0]
+        idx = np.argsort(indices, axis=1)
+        matrix = sparse.csr_matrix((size, size), dtype="complex128")
+        matrix.indices = np.take_along_axis(indices, idx, axis=1).ravel()
+        matrix.data = np.take_along_axis(data, idx, axis=1).ravel()
+        matrix.indptr = _cached_arange(size + 1) * indices.shape[1]
```

Comment:
```suggestion
        num_entries_per_row = indices.shape[1]
        matrix.indptr = _cached_arange(size + 1) *  num_entries_per_row
```

Might help make this easier to understand.